Author: Ye Shanshan<yeshanshan@uniontech.com>
Date: Thu Jul 17 10:00:00 2026
Subject: set lastInputDevice on pointer and keyboard enter for popup grab
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/753734

---

Index: qt6-wayland/src/client/qwaylandinputdevice.cpp
===================================================================
--- qt6-wayland.orig/src/client/qwaylandinputdevice.cpp
+++ qt6-wayland/src/client/qwaylandinputdevice.cpp
@@ -708,6 +708,11 @@ void QWaylandInputDevice::Pointer::point
                                 << "attempting to work around it by invalidating the current focus";
         invalidateFocus();
     }
+
+    if (!mParent->mQDisplay->lastInputDevice()) {
+        mParent->mQDisplay->setLastInputDevice(mParent, serial, window);
+    }
+
     mFocus = window->waylandSurface();
     connect(mFocus.data(), &QObject::destroyed, this, &Pointer::handleFocusDestroyed);
 
@@ -1267,9 +1272,8 @@ void QWaylandInputDevice::Keyboard::keyb
 #endif
 }
 
-void QWaylandInputDevice::Keyboard::keyboard_enter(uint32_t time, struct wl_surface *surface, struct wl_array *keys)
+void QWaylandInputDevice::Keyboard::keyboard_enter(uint32_t serial, struct wl_surface *surface, struct wl_array *keys)
 {
-    Q_UNUSED(time);
     Q_UNUSED(keys);
 
     if (!surface) {
@@ -1287,6 +1291,9 @@ void QWaylandInputDevice::Keyboard::keyb
         disconnect(mFocus, &QWaylandSurface::destroyed, this, &Keyboard::handleFocusDestroyed);
     }
 
+    if (!mParent->mQDisplay->lastInputDevice())
+        mParent->mQDisplay->setLastInputDevice(mParent, serial, window);
+
     mFocus = window->waylandSurface();
     connect(mFocus, &QWaylandSurface::destroyed, this, &Keyboard::handleFocusDestroyed);
 
Index: qt6-wayland/src/client/qwaylandinputdevice_p.h
===================================================================
--- qt6-wayland.orig/src/client/qwaylandinputdevice_p.h
+++ qt6-wayland/src/client/qwaylandinputdevice_p.h
@@ -217,7 +217,7 @@ public:
     void keyboard_keymap(uint32_t format,
                          int32_t fd,
                          uint32_t size) override;
-    void keyboard_enter(uint32_t time,
+    void keyboard_enter(uint32_t serial,
                         struct wl_surface *surface,
                         struct wl_array *keys) override;
     void keyboard_leave(uint32_t time,
Index: qt6-wayland/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
===================================================================
--- qt6-wayland.orig/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
+++ qt6-wayland/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
@@ -314,7 +314,7 @@ QWaylandXdgSurface::QWaylandXdgSurface(Q
 
     if (type == Qt::ToolTip && transientParent) {
         setPopup(transientParent);
-    } else if (type == Qt::Popup && transientParent && display->lastInputDevice()) {
+    } else if (type == Qt::Popup) {
         setGrabPopup(transientParent, display->lastInputDevice(), display->lastInputSerial());
     } else {
         setToplevel();
